### ISSUE TITLE
Calling resize_min on an empty vector causes runtime error

### ISSUE THREAD

**Issue by dahlbaek** [igraph/igraph#1479]
- [x] This issue is for the C core of igraph.
- [x] This issue is a bug report or a feature request, not a support question.

I'm not sure if this is in fact the intended behaviour, but when I try to call resize_min on an empty vector, the result is a runtime error. I'm not that strong in C, but as far as I can tell it happens because realloc called with size 0 will deallocate the array powering the vector:

https://github.com/igraph/igraph/blob/master/src/vector.pmt#L946-L963

Would it be reasonable to resize the capacity to 1 in case the vector is empty? I.e. change this implementation to something like

```
int FUNCTION(igraph_vector, resize_min)(TYPE(igraph_vector)*v) {
    size_t size;
    size_t capacity;
    BASE *tmp;
    if (v->stor_end == v->end) {
        return 0;
    }

    size = capacity = (size_t) (v->end - v->stor_begin);
    if (capacity == 0) {
        capacity = 1;
    }
    tmp = igraph_Realloc(v->stor_begin, capacity, BASE);
    if (tmp == 0) {
        IGRAPH_ERROR("cannot resize vector", IGRAPH_ENOMEM);
    } else {
        v->stor_begin = tmp;
        v->end = tmp + size;
        v->stor_end = tmp + capacity;
    }

    return 0;
}
```

**Comment by szhorvat** [comment]
Yes, I agree that realloc_min should work for zero-size vectors. I'll look into it.

It seems that we cannot count on any realloc behaviour for the zero-size case:

https://en.cppreference.com/w/c/memory/realloc

> If new_size is zero, the behavior is implementation defined (null pointer may be returned (**in which case the old memory block may or may not be freed**), or some non-null pointer may be returned that may not be used to access storage).

**Comment by vtraag** [comment]
Similar problems may actually arise with simply allocating a zero-length array using `malloc`. This is supported as an extension by `gcc`, but I believe it is not part of any C standard (although I can't find an exact references at the moment). I believe we ran into this issue of zero-length arrays in some other issue as well, but I can't recall exactly where. @szhorvat, perhaps you should then also make sure the initial capacity is always at least 1? Perhaps there are also some other cases in `vector.pmt` for which a minimum capacity of 1 should be ensured.

**Comment by szhorvat** [comment]
Yes, and not only vector, but some other data structures as well.

**Comment by vtraag** [comment]
By the way, this is the solution taken by suitesparse:

```
void *cs_malloc (CS_INT n, size_t size)
{
return (malloc (CS_MAX (n,1) * size)) ;
}

```

With similar things for `calloc` and `realloc`. Perhaps we could do something in this direction? This would ensure that regardless of the types, we would never have problems with arrays of size 0.

**Comment by szhorvat** [comment]
Yes, I was thinking along these lines. There are already macros in igraph for all relevant allocation functions—we simply need to change them to always allocate at least one byte (or unit), and ensure that these macros are always used (most of the time they are).

**Comment by szhorvat** [comment]
Normally, all allocation in igraph uses `igraph_Calloc` or `igraph_Realloc` in `igraph_memory.h`, so those could be changed to ensure that at least _some_ memory is always allocated.

**Comment by GroteGnoom** [comment]
If I want to use the current macro's of igraph_Calloc and igraph_Realloc and add some ternary conditional operator, this will cause a double evaluation. If I want to prevent that, I have to assign to a variable and then I can't return a value anymore in a platform-independent way. Can igraph_Calloc and igraph_Realloc be changed to functions? I don't see a reason why they should be macro's.

**Comment by vtraag** [comment]
I remember there were some discussions about this, with @szhorvat also adding an `igraph_malloc` function instead of a macro in https://github.com/igraph/igraph/pull/933. The argument there is that `free` and `malloc` should always come in pairs that match with the runtime. The reason that there is (also) a macro is that using the macro ensures that the pointer will be set to `NULL`, while this cannot be done in the function (at least not without complicating things even more).

The solution you added in https://github.com/GroteGnoom/igraph/commit/ca0ac66ce55ef23341330cda5fa2d796a1901eee by just adding always 1 is simple enough, and I think it should never raise any issues.

**Comment by vtraag** [comment]
Perhaps for the sake of consistency, it would be nice to have all pairs of `malloc`, `calloc` and `realloc`, `free` in both macros and functions.

**Comment by GroteGnoom** [comment]
> I remember there were some discussions about this, with @szhorvat also adding an `igraph_malloc` function instead of a macro in #933. The argument there is that `free` and `malloc` should always come in pairs that match with the runtime. The reason that there is (also) a macro is that using the macro ensures that the pointer will be set to `NULL`, while this cannot be done in the function (at least not without complicating things even more).

Aha! That clears things up. 
 
> The solution you added in [GroteGnoom@ca0ac66](https://github.com/GroteGnoom/igraph/commit/ca0ac66ce55ef23341330cda5fa2d796a1901eee) by just adding always 1 is simple enough, and I think it should never raise any issues.

The biggest concern I guess is that with data structures where lots of single nodes are allocated, like the igraph trie, there will be about 50% more memory usage.

> Perhaps for the sake of consistency, it would be nice to have all pairs of malloc, calloc and realloc, free in both macros and functions.

Yes, if I don't want to change the interface, and keep the capitalization consistent, the igraph_Calloc could call igraph_calloc, which then takes care of the always nonzero allocations. Or igraph_Calloc can call an internal function, I think I prefer that.
A downside to having public functions and macros is that it can get confusing. I see that in practice igraph_free is used to pass on to IGRAPH_FINALLY, and otherwise igraph_Free is used. But this is not completely consistent and they don't do exactly the same thing because igraph_free does not set the pointer to NULL, it only sets a copy of the pointer to NULL. This wouldn't directly be a problem for calloc, but people will still be wondering which one to choose.

**Comment by vtraag** [comment]
> The biggest concern I guess is that with data structures where lots of single nodes are allocated, like the igraph trie, there will be about 50% more memory usage.

Yes, this is true. I am not sure how often this will be the case though. Perhaps it is worth looking into using a ternary operator still. Could you clarify what you meant by the "double evaluation" that you mentioned being problematic?

>  A downside to having public functions and macros is that it can get confusing. I see that in practice igraph_free is used to pass on to IGRAPH_FINALLY, and otherwise igraph_Free is used. But this is not completely consistent and they don't do exactly the same thing because igraph_free does not set the pointer to NULL, it only sets a copy of the pointer to NULL. This wouldn't directly be a problem for calloc, but people will still be wondering which one to choose.

Yes, this is true, it should be properly documented. Perhaps for the allocation functions, we actually don't need a macro. At least, I don't immediately see the point in having macros for that instead of functions. 

The situation for `igraph_free` and `igraph_Free` is a bit special though. It is preferable to use `igraph_Free` since that also ensures the pointer is being set to `NULL`. However, we cannot pass a macro to the `IGRAPH_FINALLY` handler, because it needs to push a function pointer to the `FINALLY` stack. That is why we need to have both a function and a macro. In the end, it does not matter that much which is used, since both free the memory correctly. However, `igraph_Free` may lead to less problems preventing us from (accidentally) freeing memory twice. 

That being said, I think that almost all other macros in `igraph` are written in full capitals. Perhaps changing the name of `igraph_Free` to `IGRAPH_FREE`, might clarify that it is a macro, not a function.

So, all in all, perhaps it would be best to have:
- only allocation functions, not macros
- have both `igraph_free` function and `IGRAPH_FREE` macro.

@szhorvat, what do you think?

**Comment by GroteGnoom** [comment]
> Could you clarify what you meant by the "double evaluation" that you mentioned being problematic?

It's this problem:
https://stackoverflow.com/questions/39439181/what-is-double-evaluation-and-why-should-it-be-avoided
It occurs if you want to use a macro to first evaluate `n` to see if it is 0, and then possibly use `n` again if it is not 0. If someone then calls this macro with `igraph_Calloc(do_lots_of_things_and_return_a_number())`, then this function that returns `n` will be called twice. This shouldn't happen a lot but can be really confusing when it occurs.

**Comment by vtraag** [comment]
OK, thanks for the explanation! So this would be an argument in favour of only having allocation functions, not macros, in which case this should not be a problem, correct? We could then still have a sort of `MAX` macro, if we want.

**Comment by szhorvat** [comment]
I'll read the thread in detail and chime in more over the weekend, I just wanted to say that:

 - I am not concerned about double evaluation. These are internal functions/macros that we know how to use correctly. Also, it will usually happen only when there's a function call, e.g. `#define MAX((a),(b)) ((a > b) ? (a) : (b))` will have no problem with `MAX(x+y, z)`: the compiler is smart enough not to calculate x+y twice. It's only a potential problem with `MAX(f(x), f(y))`, and even then only if the compiler can't see into what `f()` is, or `f()` has explicit side effects.

 - There are things that only macros can do, such as `igraph_Free` setting the variable to `NULL` or `igraph_Calloc` determining the `sizeof` the the type.

 - Function calls will have more overhead than macro (macros are essentially inline), but given how slow a `calloc` is likely to be anyway, this almost certainly doesn't matter.

 - We only really need a `max` operation here, so if we do go the function call route (which I think is not actually necessary—we can take a bit of care with `igraph_Calloc`), then only `max` needs to be implemented as a function call, the rest can stay as a macro.

**Comment by vtraag** [comment]
>     * `igraph_Calloc` determining the `sizeof` the the type.

Right, this is a good point, I had not yet considered that. In `suitesparse` the approach that I mentioned [here](https://github.com/igraph/igraph/issues/1479#issuecomment-716055638) indeed requires one to always specify `sizeof( [type] )` on the part of the caller. Whether that is really problematic could be open to discussion of course.

The double evaluation does sound like a potentially thorny issue to me. Usually this will not be a problem, but it might get ugly in certain cases. In that case, it might be easier to have more robust methods that we don't have to care about these considerations, but that requires us to explicitly write `sizeof( [type] )`, than to be aware of possible pitfalls when using the method.

Anyway, perhaps we should see in which cases this double evaluation is really problematic.

**Comment by ntamas** [comment]
In general I agree with @szhorvat .

> The double evaluation does sound like a potentially thorny issue to me. Usually this will not be a problem, but it might get ugly in certain cases.

I think that C programmers working with macros (_any_ kind of macros) should be aware of the double evaluation problem as it is related to macros in general, not only to the type of macros we define here. All that we need is to make it damn clear that `igraph_Calloc`, `igraph_Free` et al are macros, and making them `ALL_CAPS_UNDERSCORE_CASE` should do the trick.

**Comment by vtraag** [comment]
> I think that C programmers working with macros (_any_ kind of macros) should be aware of the double evaluation problem as it is related to macros in general, not only to the type of macros we define here.

That is true of course, but we are making the situation more complex by nesting macros.

> All that we need is to make it damn clear that `igraph_Calloc`, `igraph_Free` et al are macros, and making them `ALL_CAPS_UNDERSCORE_CASE` should do the trick.

I think this should indeed be done.

**Comment by szhorvat** [comment]
> That is true of course, but we are making the situation more complex by nesting macros

It is not necessary to nest macros. A `? :` operator will do.

**Comment by szhorvat** [comment]
I went through (hopefully) all uses of `igraph_Calloc` and `igraph_Realloc` and checked if they are ever invoked with a function call, which could _potentially_ lead to double evaluation. I fixed all of these. There were only a handful. See 6ac0b0f0fcdeeba127cccb183851d2b2ddbef906  Note that if these were to result in double evaluation, the only ill effect would be a _potential_ but still negligible performance degradation. All functions were size-measuring ones, such as `igraph_ecount`, `vector_size`, `matrix_nrows`, `strlen`, i.e. having no side effects. I would even expect the compiler to know that `strlen` has no size effects and optimize away any double evaluation. As for the rest, an `ecount` has minimal performance impact, and it is not uncommon that it is "unnecessarily" called multiple times.

In summary:

 - Due to the nature of these functions, there is a low risk that they will be misused in a way that leads to double evaluation
 - If they are, the effects are very likely to be negligible

**Comment by szhorvat** [comment]
As an experiment, I change `igraph_Calloc/Realloc` like this:

```
#define igraph_Calloc(n,t)    (t*) calloc( (n) > 0 ? (size_t)(n) : (size_t)1, sizeof(t) )
#define igraph_Realloc(p,n,t) (t*) realloc((void*)(p), (n) > 0 ? (size_t)((n)*sizeof(t)) : (size_t)1)
```

I can confirm that this makes all "Use of zero-allocated memory" Clang Static Analyzer warnings go away.

I should not that when I was looking at individual usages of `igraph_Calloc/Realloc`, I did notice that _some_ of code related to fundamental data structures (vector, stack, etc.) already had protection against zero-allocation, either by adding 1 or by checking for zero explicitly. It would be nice to find and remove these.

**Comment by szhorvat** [comment]
> I remember there were some discussions about this, with @szhorvat also adding an `igraph_malloc` function instead of a macro in #933. The argument there is that `free` and `malloc` should always come in pairs that match with the runtime. The reason that there is (also) a macro is that using the macro ensures that the pointer will be set to `NULL`, while this cannot be done in the function (at least not without complicating things even more).

I just wanted to note that the motivation for `igraph_malloc` was entirely different:  `igraph_Calloc/Realloc` are really for internal use (and IMO we should eventually remove them from public headers). `igraph_malloc` is specifically for public use, to ensure that when the user allocates memory, but igraph itself frees it, it is possible to use a matching pair of `malloc`/`free` functions. One such example is when allocating elements for a `vector_ptr` manually, but freeing them with `vector_ptr_free/destroy`.

I actually had a need for `igraph_malloc` for the Mathematica interface. In the past, I could only compile igraph (with the features I needed) using MSYS2. But for various reasons, I used MSVC for compiling the glue code between Mathematica and igraph. If I allocated with MSVC's `malloc`, but let igraph free the memory with MSYS2's `free`, things would go very wrong (usually the program would crash).

The motivation for `igraph_free` is similar, with one additional internal use case: `IGRAPH_FINALLY` requires a function, not a macro. Thus we needed the function `igraph_free` in addition to the `igraph_Free` macro.

**Comment by szhorvat** [comment]
I would vote for:

 - Capitalizing `IGRAPH_FREE` and `IGRAPH_CALLOC`. These are in public headers, but they are really for internal use, so I'm fine with changing them on master.
 - Use the macro definitions from my post above: https://github.com/igraph/igraph/issues/1479#issuecomment-791909912
 - Keep them as macros, not functions.
 - Skim through the present uses if `igraph_Calloc/Realloc` and remove any existing mitigations against zero-allocation.

**Comment by vtraag** [comment]
@szhorvat sounds good! Could we then also remove `IGRAPH_FREE` and `IGRAPH_CALLOC` from the public headers and move them to an internal header

 You would also include `IGRAPH_CALLOC` and `IGRAPH_REALLOC` I presume, by the way?

**Comment by szhorvat** [comment]
> Could we then also remove `IGRAPH_FREE` and `IGRAPH_CALLOC` from the public headers and move them to an internal header?

I thought some more about this, and I think we should in fact not do this.  I would keep them public but undocumented.

I would like to make it possible to implement most new functionality using only things exposed through the public header. I sometime implement new functions within the Mathematica interface, with the intention to move it to the C core later. There is value in being able to use those "internal" features which are used in almost all functions, such as the `FINALLY_...` stuff, error handling, as memory allocation. In fact I already find it annoying that I cannot easily use `IGRAPH_ALLOW_INTERRUPTION()` like this.

**Comment by ntamas** [comment]
This will be fixed by #1726. I'll push a regression test for this issue in #1726 so we don't get bitten by this again.
